src/rgw/rgw_op.cc

```diff
@@ -5949,6 +5949,39 @@ void RGWCompleteMultipart::execute(optional_yield y)
   op_ret = serializer->try_lock(dur, y);
   if (op_ret < 0) {
     ldpp_dout(this, 0) << "failed to acquire lock" << dendl;
+    if (op_ret == -ENOENT) {
+      std::string oid = mp.get_part(1);
```

Comment:
i don't think it's right to test the existence of the first part object. that will exist even before multipart completes, right?

instead we should be checking the head object, which is created by `write_meta()` at the end of `execute()`. and since we might be overwriting an object created earlier, we should compare that head object's manifest with the `CompleteMultipartUpload` payload (an xml document that has the size and etag of each part) to make sure it's the same upload

Reply:
you might refer to rgw_etag_verifier.* for an example that a) decodes the RGWObjManifest from the head object's RGW_ATTR_MANIFEST xattr, b) checks whether the manifest is multipart or 'atomic', and c) calculates the multipart ETag from its component parts. you can compare that composite ETag with the head object's RGW_ATTR_ETAG xattr

Reply:
I'm going to replace the static cast bits before a new backend goes in, as part of the refactoring work I'm doing now, so I don't think it makes much difference.

Reply:
Added a note to the tracker [1] showing the Quay complete-miltipart flow.
will look into the ETag re-calculation next.

[1] https://tracker.ceph.com/issues/50141#note-2

Reply:
- moved into a helper function.
- switched to a dynamic_cast and added a TODO
- instead of checking for only the 1st part, checking for the existence of all the parts numbers from fileparts json/xml
- regarding testing the existence of object, concerning a false-positive in which they will exist even before multipart complete, on the 1st completion https://github.com/ceph/ceph/blob/d616f353bde9a2037a49aeace41ffb8944bb17cb/src/rgw/rgw_op.cc#L5949
  does not return an error and the code flows as previously, the new check for the parts only occurs on consecutive completions. (the flow detailed at https://tracker.ceph.com/issues/50141#note-2)